docs/cheat-sheets/java-xxe.mdx

```diff
@@ -0,0 +1,1817 @@
+---
+slug: java-xxe
+toc_max_heading_level: 4
+---
+
+# XML External Entity prevention for Java
+
+This is an XML External Entity (XXE) prevention cheat sheet by r2c. It contains code patterns of potential XXE in an application. Instead of scrutinizing code for exploitable vulnerabilities, the recommendations in this cheat sheet pave a safe road for developers that mitigates the possibility of XXE in your code. By 
+following these recommendations, you can be reasonably sure your code is free of XXE.
+
+## Mitigation Summary
+
+XML security in Java is not easy to summarize. On the one hand, the XML standard allows multiple ways to include external content, including but not limited to external Document Type Definitions (DTD), external Entity References to external data, general entity references, and external parameter references.
+There are also several ways to include external content in XML Schemas and stylesheets.
+
+On the other hand, the Java API for XML Processing (JAXP) contains multiple interfaces to process XML content. The Document Object Model (DOM) interface, the Simple API for XML (SAX) interface, and the Streaming API for XML (StAX) interface.
+Each of these interfaces has its mitigation measures to disable allowing the use of different sources of external content.
+The inconsistency in availability and even the effect of these measures across the different APIs, combined with the variety of ways to include external content, makes it difficult to make a general recommendation. Many configurations that are secure for one parser are not secure for another.
+However, the following settings have a consistent effect on a subset of parsers
+
+- For `DocumentBuilderFactory`. `SAXBuilder`, `SAXParser`, `SAXReader`, and `XMLReader` use:
+    ```
+    setFeature(http://apache.org/xml/features/disallow-doctype-decl, true);
+    ```
+- For `SAXTransformer`, `SchemaFactory`, and `Validator` use:
+    ```
+    setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    ```
+
+The complete Semgrep ruleset for this cheatsheet is available as part of the rules available in the Team tier. This ruleset includes rules to verify these settings as well as several other combinations that are secure.
```

Comment:
@0xDC0DE Please, include a link to this ruleset. Only users of the Team tier will be able to access it.